### PR TITLE
1D test for Reduce layer

### PR DESCRIPTION
**Pull request by Abdurrahheem**
This PR introduces test for `Reduce` layer to test its functionality for 1D arrays

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by Abdurrahheem** [comment]
@fengyuentau. I get strange result for `log_sum` reduce operation. Expected result should be `0` for 1D vector of `1's`. The actually result differs from that. Could you please take a look on it?

**Comment by Abdurrahheem** [comment]
This #25103 issue is also relevant

**Comment by Abdurrahheem** [comment]
@fengyuentau

**Comment by Abdurrahheem** [comment]
@dkurt could you please take a look?

File: modules/dnn/test/test_layers_1d.cpp [hunk 1]
```diff
@@ -245,4 +245,80 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Elemwise_1d_Test, Combine(
...
+            }
+        }
+        if (operation == "mean") {
+            result = sum / input.total();
+        } else if (operation == "l2") {
+            result = std::sqrt(result);
+        } else if (operation == "log_sum") {
+            for (int i = 0; i < input.total(); ++i) {
+                result += std::log(input.at<float>(i));
+            }
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:281]: Why not as a part of the previous loop?

File: modules/dnn/test/test_layers_1d.cpp [hunk 2]
```diff
@@ -245,4 +245,80 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Elemwise_1d_Test, Combine(
...
+
+    std::vector<Mat> inputs{input};
+    std::vector<Mat> outputs;
+
+    runLayer(layer, inputs, outputs);
+    ASSERT_EQ(shape(output_ref), shape(outputs[0]));
+    normAssert(output_ref, outputs[0]);
+}
+INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Reduce_Test, Combine(
+/*input blob shape*/    Values(0, 1),
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:319]: I recommend use shapes directly. Including 1D vector with multiple values.

File: modules/dnn/test/test_layers_1d.cpp [hunk 3]
```diff
@@ -245,4 +245,84 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Elemwise_1d_Test, Combine(
...
+    std::vector<Mat> outputs;
+
+    runLayer(layer, inputs, outputs);
+    ASSERT_EQ(shape(output_ref), shape(outputs[0]));
+    normAssert(output_ref, outputs[0]);
+}
+INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Reduce_Test, Combine(
+/*input blob shape*/    Values(
+    std::make_tuple(0, std::vector<int>{0}),
+    std::make_tuple(1, std::vector<int>{1}),
```
Comment (fengyuentau) [inline modules/dnn/test/test_layers_1d.cpp:321]: Could you define reduction operation on 0d / 1d tensor? Any reference?

Comment (Abdurrahheem) [reply modules/dnn/test/test_layers_1d.cpp:321]: formally there is not any. This test is to check if layer fails on singleton matrices

Comment (fengyuentau) [reply modules/dnn/test/test_layers_1d.cpp]: There are some in onnxruntime: https://github.com/microsoft/onnxruntime/blob/4a196d15940b0f328735c888e2e861d67602ffcf/onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc#L326 (Search for 0D).

File: modules/dnn/test/test_layers_1d.cpp [hunk 4]
```diff
@@ -349,4 +349,83 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Expand_Test, Combine(
...
+    ASSERT_EQ(shape(output_ref), shape(outputs[0]));
+    normAssert(output_ref, outputs[0]);
+}
+INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Reduce_Test, Combine(
+/*input blob shape*/    Values(
+    std::vector<int>({}),
+    std::vector<int>({1}),
+    std::vector<int>({1, 4})
+    ),
+/*input blob shape*/    Values("max", "min", "mean", "sum", "sum_square", "l1", "l2", "prod", "log_sum", "log_sum_exp"))
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:427]: incorrect comment

File: modules/dnn/test/test_layers_1d.cpp [hunk 5]
```diff
@@ -349,4 +349,83 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Expand_Test, Combine(
...
+            }
+        }
+        return result;
+    };
+
+    std::vector<int> input_shape = get<0>(GetParam());
+    std::string reduce_operation = get<1>(GetParam());
+
+    if (input_shape.size() == 2 && reduce_operation == "log_sum") // both output and reference are nans
+        return;
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:397]: I guess we should skip it earlier

File: modules/dnn/test/test_layers_1d.cpp [hunk 6]
```diff
@@ -349,4 +349,83 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Expand_Test, Combine(
...
+    lp.set("axis", 0);
+    lp.set("keepdims", true);
+    Ptr<ReduceLayer> layer = ReduceLayer::create(lp);
+
+    cv::Mat input(input_shape.size(), input_shape.data(), CV_32F, 1.0);
+    cv::randn(input, 0.0, 1.0);
+
+    float out_value = reduceOperation(input, reduce_operation);
+
+    cv::Mat output_ref = cv::Mat(input_shape.size(), (input_shape.size() <= 1) ? input_shape.data() : std::vector<int>({1, 1}).data(), CV_32F, out_value);
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:412]:
```suggestion
    cv::Mat output_ref = cv::Mat(std::vector<int>(input_shape.size(), 1), CV_32F, out_value);
```

File: modules/dnn/test/test_layers_1d.cpp [hunk 7]
```diff
@@ -480,4 +480,84 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Scatter_Test, Combine(
...
+    lp.name = "reduceLayer";
+    lp.set("reduce", reduce_operation);
+    lp.set("axis", 0);
+    lp.set("keepdims", true);
+    Ptr<ReduceLayer> layer = ReduceLayer::create(lp);
+
+    cv::Mat input(input_shape.size(), input_shape.data(), CV_32F, 1.0);
+    cv::randn(input, 0.0, 1.0);
+
+    float out_value = reduceOperation(input, reduce_operation);
```
Comment (opencv-alalek) [inline modules/dnn/test/test_layers_1d.cpp:541]: Why do still we have ONE output value for matrix?

Don't write WEAK tests.

Comment (Abdurrahheem) [reply modules/dnn/test/test_layers_1d.cpp:541]: Sorry, I do not understand you. What could be output for 1D(or 0D) matrix under reduction operation other than one output  value? What do you mean my output value?

File: modules/dnn/test/test_layers_1d.cpp (hunk 1 above)
Comment (Abdurrahheem) [reply modules/dnn/test/test_layers_1d.cpp]: Done! (did not see the comment, sorry)

**Comment by Abdurrahheem** [comment]
@opencv-alalek anything else?

File: modules/dnn/test/test_layers_1d.cpp (hunk 7 above)
Comment (dkurt) [reply modules/dnn/test/test_layers_1d.cpp:538]: in case of `[4, 4]` input and `axis=1` output is `[4, 1]`, so it also can be tested

**Comment by asmorkalov** [comment]
@dkurt @opencv-alalek Please take a look again.

**Comment by asmorkalov** [comment]
```
[  FAILED  ] Layer_Reduce_Test.Accuracy_01D/8, where GetParam() = ({}, "log_sum")
[  FAILED  ] Layer_Reduce_Test.Accuracy_01D/18, where GetParam() = ({ 1 }, "log_sum")
[  FAILED  ] Layer_Reduce_Test.Accuracy_01D/19, where GetParam() = ({ 1 }, "log_sum_exp")
```

File: modules/dnn/test/test_layers_1d.cpp [hunk 8]
```diff
@@ -480,6 +480,116 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Scatter_Test, Combine(
...
 
 
+typedef testing::TestWithParam<tuple<std::vector<int>, std::string>> Layer_Reduce_Test;
+TEST_P(Layer_Reduce_Test, Accuracy_01D)
+{
+    auto reduceOperation = [](const cv::Mat& input, const std::string& operation, int axis) -> cv::Mat {
+        // Initialize result matrix
+        cv::Mat result;
+        if (shape(input).size() == 0 || shape(input).size() == 1){
+            result = cv::Mat(shape(input).size(), shape(input).data(), CV_32F);
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:490]: Might be incorrect for `std::vector<int>({4})` test case (not represented). Number of input dimensions is 1 however output shape is not equal to input shape.

File: modules/dnn/test/test_layers_1d.cpp [hunk 9]
```diff
@@ -480,6 +480,116 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Scatter_Test, Combine(
...
+        } else {
+            if (axis == 0) {
+                result = cv::Mat::zeros(1, input.cols, CV_32F);
+            } else {
+                result = cv::Mat::zeros(input.rows, 1, CV_32F);
+            }
+        }
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:497]: Can we simplify to something like this?
```cpp
MatShape shape(input).
if (!shape.empty())
  shape[axis] = 1;
result = cv::Mat::zeros(shape.size(), &shape[0], CV_32F);
```

File: modules/dnn/test/test_layers_1d.cpp [hunk 10]
```diff
@@ -480,6 +480,116 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Scatter_Test, Combine(
...
+                    if (operation == "sum" || operation == "l1" || operation == "l2"
+                        || operation == "sum_square" || operation == "mean" || operation == "log_sum"
+                        || operation == "log_sum_exp") res = 0;
+                    else if (operation == "prod") res = 1;
+                }
+
+                if (operation == "sum" || operation == "mean") res += value;
+                else if (operation == "sum_square") {
+                    if (shape(input).size() == 2 && shape(input)[0] == 1)
+                        res += value;
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:515]: Please clarify this condition. For example, `{1, 4}` input, axis=1 should be reduced to `{1x1}`:
 `[[a, b, c, d]]` -> `[[a*a + b*b + c*c + d*d]]`

**Comment by asmorkalov** [comment]
@fengyuentau @dkurt Could you take a look again?

File: modules/dnn/test/test_layers_1d.cpp [hunk 11]
```diff
@@ -480,6 +480,133 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Scatter_Test, Combine(
...
+                        || operation == "log_sum_exp") res = 0;
+                    else if (operation == "prod") res = 1;
+                }
+
+                if (operation == "sum" || operation == "mean") res += value;
+                else if (operation == "sum_square") {
+                    if (shape(input).size() == 2 && shape(input)[0] == 1 && axis==0)
+                        res += value;
+                    else if (shape(input).size() == 2 && shape(input)[1] == 1 && axis==1)
+                        res += value;
```
Comment (dkurt) [inline modules/dnn/test/test_layers_1d.cpp:522]: Should be square in both cases:
```python
@ost.script()
def layer_sum_square(x: ost.FLOAT[1, 4]) -> ost.FLOAT[1, 4]:
    return opset18.ReduceSumSquare(x, axes=[0], keepdims=1)

make_model_and_data(layer_sum_square, np.random.rand(1, 4).astype(np.float32))
```
```
input [[0.5488135  0.71518934 0.60276335 0.5448832 ]]
output [[0.30119628 0.51149577 0.36332366 0.29689768]]
```


```python
@ost.script()
def layer_sum_square(x: ost.FLOAT[1, 4]) -> ost.FLOAT[1, 4]:
    return opset18.ReduceSumSquare(x, axes=[0], keepdims=1)

make_model_and_data(layer_sum_square, np.random.rand(1, 4).astype(np.float32))
```
```
input [[0.5488135  0.71518934 0.60276335 0.5448832 ]]
output [[0.30119628 0.51149577 0.36332366 0.29689768]]
```

```python
@ost.script()
def layer_sum_square(x: ost.FLOAT[4, 1]) -> ost.FLOAT[4, 1]:
    return opset18.ReduceSumSquare(x, axes=[1], keepdims=1)

make_model_and_data(layer_sum_square, np.random.rand(4, 1).astype(np.float32))
```
```
input [[0.5488135 ]
 [0.71518934]
 [0.60276335]
 [0.5448832 ]]
output [[0.30119628]
 [0.51149577]
 [0.36332366]
 [0.29689768]]
 ```
 source: https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/onnx/generate_onnx_models_with_onnxscript.py

File: modules/dnn/test/test_layers_1d.cpp [hunk 12]
```diff
@@ -480,6 +480,133 @@ INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Scatter_Test, Combine(
...
+    std::vector<Mat> outputs;
+
+    runLayer(layer, inputs, outputs);
+    ASSERT_EQ(outputs.size(), 1);
+    ASSERT_EQ(shape(output_ref), shape(outputs[0]));
+    normAssert(output_ref, outputs[0]);
+}
+INSTANTIATE_TEST_CASE_P(/*nothing*/, Layer_Reduce_Test, Combine(
+/*input blob shape*/    Values(
+    std::vector<int>({}),
```
Comment (fengyuentau) [inline modules/dnn/test/test_layers_1d.cpp:597]: nit: use without parenthesis for simplificity, e.g. `std::vector<int>{}`.

File: modules/dnn/test/test_layers_1d.cpp [hunk 13]
```diff
@@ -586,8 +581,12 @@ TEST_P(Layer_Reduce_Test, Accuracy_01D)
     cv::Mat output_ref = reduceOperation(input, reduce_operation, axis);
     std::vector<Mat> inputs{input};
     std::vector<Mat> outputs;
+    std::cout << "input: " << input << std::endl;
 
     runLayer(layer, inputs, outputs);
+    std::cout << "shape of output: " << shape(outputs[0]) << std::endl;
+    std::cout << "output: " << outputs[0] << std::endl;
+    std::cout << "ref: " << output_ref << std::endl;
```
Comment (asmorkalov) [inline modules/dnn/test/test_layers_1d.cpp:589]: Please drop debug printing.

**Comment by asmorkalov** [comment]
@fengyuentau @dkurt Could you take a look again?

**Review by dkurt — APPROVED** [review]

**Review by fengyuentau — APPROVED** [review]

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**